File: android/app/src/main/kotlin/com/fatwabot/app/navigation/RemembranceScreen.kt
package com.fatwabot.app.navigation

import androidx.activity.compose.BackHandler
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.LazyListState
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.Clear
import androidx.compose.material.icons.filled.Favorite
import androidx.compose.material.icons.filled.FavoriteBorder
import androidx.compose.material.icons.filled.MenuBook
import androidx.compose.material.icons.filled.PlayArrow
import androidx.compose.material.icons.filled.Search
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextField
import androidx.compose.material3.TextFieldDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.fatwabot.app.R
import com.fatwabot.core.common.ContentFocus
import com.fatwabot.core.content.AzkarCategory
import com.fatwabot.core.content.AzkarItem
import com.fatwabot.core.content.Dua
import com.fatwabot.core.content.DuaCategory
import com.fatwabot.core.designsystem.ArabicContentCard
import com.fatwabot.core.designsystem.BrandEmptyState
import com.fatwabot.core.designsystem.ColorTokens
import com.fatwabot.core.designsystem.DarkTokens
import com.fatwabot.core.designsystem.LightTokens
import com.fatwabot.core.designsystem.RepeatCountLabel
import com.fatwabot.core.designsystem.brandScreenBackground
import com.fatwabot.feature.azkar.AzkarSearch
import com.fatwabot.feature.azkar.AzkarSessionScreen
import com.fatwabot.feature.azkar.AzkarViewModel
import com.fatwabot.feature.dua.DuaViewModel
import java.util.Locale
import kotlinx.coroutines.delay
import com.fatwabot.feature.azkar.R as AzkarR
import com.fatwabot.feature.dua.R as DuaR

/** Which library the merged screen opens on — set by which deep link/tile got
 *  here (`fatwabot://azkar` vs `fatwabot://dua`). Both land on this one
 *  screen; this only picks the initially-selected chip. */
enum class RemembranceSegment { AZKAR, DUA }

private sealed class RemembranceSelection {
    data class Azkar(val id: String) : RemembranceSelection()
    data class Dua(val id: String) : RemembranceSelection()
    object Favorites : RemembranceSelection()
}

/**
 * Azkar and Du'a merged into one screen behind a single category strip
 * (client direction, 2026-08-15) — replacing the earlier Azkar/Du'a segmented
 * control (2026-07-26) with one continuous row of chips spanning both
 * corpora, so the screen reads as one table of contents rather than two tabs
 * the reader has to know to switch between. Mirror of iOS `RemembranceScreen`.
 *
 * The two libraries still keep their own view models and their own session/
 * favourites logic; only the chip strip, search field and card list are
 * unified here. A chip's *kind* (azkar category, dua category, or dua
 * favourites) decides which card style and per-item actions render below it.
 */
@Composable
fun RemembranceScreen(
    initial: RemembranceSegment,
    onExit: () -> Unit,
    azkarViewModel: AzkarViewModel = hiltViewModel(),
    duaViewModel: DuaViewModel = hiltViewModel(),
    locale: String = "ar",
    /** Which specific item to scroll to, from a content-reminder tap. Only
     *  azkar reminders exist today (there is no dua notification kind), so
     *  this only ever selects an azkar chip. */
    focus: ContentFocus? = null,
) {
    val azkarState by azkarViewModel.state.collectAsStateWithLifecycle()
    val duaState by duaViewModel.state.collectAsStateWithLifecycle()
    val tokens = if (isSystemInDarkTheme()) DarkTokens else LightTokens

    var selection by remember { mutableStateOf<RemembranceSelection?>(null) }
    var query by remember { mutableStateOf("") }
    // Azkar-only push, exactly as before the merge: starting a session pushes
    // full-screen and owns the title bar until it's dismissed. Du'a no longer
    // pushes a detail — its library shows each dua in full.
    var sessionCategory by remember { mutableStateOf<AzkarCategory?>(null) }
    var highlightedItemId by remember { mutableStateOf<String?>(null) }
    val azkarListState = rememberLazyListState()

    // The counting session is a level RootScaffold can't see, so it needs its
    // own handler. Registered inside this screen, it takes priority over the
    // root one — back closes the session first, then leaves the screen.
    BackHandler(enabled = sessionCategory != null) { sessionCategory = null }

    LaunchedEffect(locale) {
        azkarViewModel.loadCategories(locale)
        duaViewModel.loadCategories(locale)
        // Land on the category a reminder's item actually belongs to —
        // without this the item may not even be in `visibleAzkarItems` yet,
        // since that is filtered to whichever chip is selected.
        val focusCategoryId = focus?.categorySlug
        if (focusCategoryId != null) {
            selection = RemembranceSelection.Azkar(focusCategoryId)
        } else if (selection == null) {
            selection = if (initial == RemembranceSegment.DUA) {
                duaViewModel.state.value.categories.firstOrNull()?.let { RemembranceSelection.Dua(it.id) }
            } else {
                azkarViewModel.state.value.categories.firstOrNull()?.let { RemembranceSelection.Azkar(it.id) }
            }
        }
    }

    val showsFavoritesChip = duaState.favoriteDuas.isNotEmpty()

    // Resolves the stored selection to something guaranteed valid, falling
    // back the same way each original screen did when the stored id goes
    // stale: a content resync dropping a category, or unfavouriting the
    // last dua while favourites is what's selected.
    val resolvedSelection: RemembranceSelection? = run {
        val current = selection
        val stillValid = when (current) {
            is RemembranceSelection.Azkar -> azkarState.categories.any { it.id == current.id }
            is RemembranceSelection.Dua -> duaState.categories.any { it.id == current.id }
            RemembranceSelection.Favorites -> showsFavoritesChip
            null -> false
        }
        if (stillValid) {
            current
        } else {
            azkarState.categories.firstOrNull()?.let { RemembranceSelection.Azkar(it.id) }
                ?: duaState.categories.firstOrNull()?.let { RemembranceSelection.Dua(it.id) }
        }
    }

    val selectedAzkarCategory = (resolvedSelection as? RemembranceSelection.Azkar)?.let { sel ->
        azkarState.categories.firstOrNull { it.id == sel.id }
    }
    val selectedDuaCategory = (resolvedSelection as? RemembranceSelection.Dua)?.let { sel ->
        duaState.categories.firstOrNull { it.id == sel.id }
    }

    // Search covers title, matn and source for azkar and is scoped to the
    // active category, unlike dua search below.
    val visibleAzkarItems = AzkarSearch.filter(selectedAzkarCategory?.items.orEmpty(), query)

    // Dua search already spans every category once a query is entered
    // (`DuaViewModel.UiState.searchResults`) — kept as-is rather than
    // narrowed to the active chip.
    val duaSearchResults = duaState.searchResults
    val visibleDuas: List<Dua> = when {
        duaSearchResults != null -> duaSearchResults
        resolvedSelection is RemembranceSelection.Favorites -> duaState.favoriteDuas
        else -> selectedDuaCategory?.duas.orEmpty()
    }

    fun select(newSelection: RemembranceSelection) {
        selection = newSelection
        // A query from the previous chip almost never applies to the next
        // one, and leaving it set shows an empty list that reads as though
        // the category itself is empty.
        query = ""
        duaViewModel.updateSearchQuery("")
    }

    WorshipDetailScaffold(
        title = sessionCategory?.name ?: stringResource(R.string.worship_remembrance),
        onBack = {
            if (sessionCategory != null) sessionCategory = null else onExit()
        },
    ) {
        if (sessionCategory != null) {
            AzkarSessionScreen(category = sessionCategory!!, viewModel = azkarViewModel)
        } else {
            Column(modifier = Modifier.fillMaxSize().brandScreenBackground(tokens)) {
                SearchField(
                    query = query,
                    tokens = tokens,
                    onChange = {
                        query = it
                        duaViewModel.updateSearchQuery(it)
                    },
                )
                CategoryChips(
                    azkarCategories = azkarState.categories,
                    duaCategories = duaState.categories,
                    showsFavoritesChip = showsFavoritesChip,
                    resolvedSelection = resolvedSelection,
                    tokens = tokens,
                    onSelect = ::select,
                )

                if (azkarState.categories.isEmpty() && duaState.categories.isEmpty()) {
                    if (!azkarState.hasLoadedCategories || !duaState.hasLoadedCategories) {
                        LoadingBox(tokens)
                    } else {
                        BrandEmptyState(
                            icon = Icons.Filled.MenuBook,
                            message = stringResource(AzkarR.string.azkar_empty_message),
                            tokens = tokens,
                        )
                    }
                } else {
                    when (resolvedSelection) {
                        is RemembranceSelection.Azkar -> AzkarContentSection(
                            category = selectedAzkarCategory,
                            items = visibleAzkarItems,
                            isCompletedToday = selectedAzkarCategory?.let { azkarViewModel.isCompletedToday(it.id) } == true,
                            tokens = tokens,
                            locale = locale,
                            listState = azkarListState,
                            focusItemId = focus?.contentId,
                            highlightedItemId = highlightedItemId,
                            onHighlightChange = { highlightedItemId = it },
                            onStartSession = { sessionCategory = it },
                        )
                        is RemembranceSelection.Dua, RemembranceSelection.Favorites -> DuaContentSection(
                            duas = visibleDuas,
                            isSearching = duaSearchResults != null,
                            isFavorite = duaState::isFavorite,
                            onToggleFavorite = duaViewModel::toggleFavorite,
                            tokens = tokens,
                        )
                        null -> LoadingBox(tokens)
                    }
                }
            }
        }
    }
}

@Composable
private fun LoadingBox(tokens: ColorTokens) {
    Box(modifier = Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
        CircularProgressIndicator(color = tokens.primary)
    }
}

@Composable
private fun SearchField(query: String, tokens: ColorTokens, onChange: (String) -> Unit) {
    TextField(
        value = query,
        onValueChange = onChange,
        singleLine = true,
        placeholder = { Text(stringResource(AzkarR.string.azkar_search_placeholder)) },
        leadingIcon = { Icon(Icons.Filled.Search, contentDescription = null) },
        trailingIcon = {
            if (query.isNotEmpty()) {
                IconButton(onClick = { onChange("") }) {
                    Icon(
                        Icons.Filled.Clear,
                        contentDescription = stringResource(AzkarR.string.azkar_search_clear),
                    )
                }
            }
        },
        shape = RoundedCornerShape(14.dp),
        colors = TextFieldDefaults.colors(
            focusedContainerColor = tokens.surfaceElevated,
            unfocusedContainerColor = tokens.surfaceElevated,
            focusedIndicatorColor = Color.Transparent,
            unfocusedIndicatorColor = Color.Transparent,
            focusedTextColor = tokens.onSurface,
            unfocusedTextColor = tokens.onSurface,
        ),
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 20.dp)
            .padding(bottom = 12.dp),
    )
}

@Composable
private fun CategoryChips(
    azkarCategories: List<AzkarCategory>,
    duaCategories: List<DuaCategory>,
    showsFavoritesChip: Boolean,
    resolvedSelection: RemembranceSelection?,
    tokens: ColorTokens,
    onSelect: (RemembranceSelection) -> Unit,
) {
    LazyRow(
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        contentPadding = PaddingValues(horizontal = 20.dp),
        modifier = Modifier.fillMaxWidth().padding(bottom = 14.dp),
    ) {
        items(azkarCategories, key = { "azkar-${it.id}" }) { category ->
            val selection = RemembranceSelection.Azkar(category.id)
            Chip(
                title = category.name,
                icon = null,
                isSelected = resolvedSelection == selection,
                tokens = tokens,
                onClick = { onSelect(selection) },
            )
        }
        if (showsFavoritesChip) {
            item(key = "favorites") {
                Chip(
                    title = stringResource(DuaR.string.dua_favorites),
                    icon = Icons.Filled.Favorite,
                    isSelected = resolvedSelection == RemembranceSelection.Favorites,
                    tokens = tokens,
                    onClick = { onSelect(RemembranceSelection.Favorites) },
                )
            }
        }
        items(duaCategories, key = { "dua-${it.id}" }) { category ->
            val selection = RemembranceSelection.Dua(category.id)
            Chip(
                title = category.name,
                icon = null,
                isSelected = resolvedSelection == selection,
                tokens = tokens,
                onClick = { onSelect(selection) },
            )
        }
    }
}

@Composable
private fun Chip(
    title: String,
    icon: ImageVector?,
    isSelected: Boolean,
    tokens: ColorTokens,
    onClick: () -> Unit,
) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(5.dp),
        modifier = Modifier
            .clip(CircleShape)
            .background(if (isSelected) tokens.primary else tokens.surfaceElevated)
            .clickable(onClick = onClick)
            .padding(horizontal = 16.dp, vertical = 9.dp),
    ) {
        if (icon != null) {
            Icon(
                icon,
                contentDescription = null,
                tint = if (isSelected) tokens.onPrimary else tokens.onSurface,
                modifier = Modifier.size(14.dp),
            )
        }
        Text(
            text = title,
            color = if (isSelected) tokens.onPrimary else tokens.onSurface,
            fontWeight = if (isSelected) FontWeight.SemiBold else FontWeight.Normal,
            style = MaterialTheme.typography.bodyMedium,
        )
    }
}

// --- Azkar content ---

/**
 * The way into the counting session, kept at the top of the list rather than
 * on every card: starting a session is a per-category action, and repeating
 * it on every card would imply each one starts its own.
 */
@Composable
private fun AzkarContentSection(
    category: AzkarCategory?,
    items: List<AzkarItem>,
    isCompletedToday: Boolean,
    tokens: ColorTokens,
    locale: String,
    listState: LazyListState,
    focusItemId: String?,
    highlightedItemId: String?,
    onHighlightChange: (String?) -> Unit,
    onStartSession: (AzkarCategory) -> Unit,
) {
    if (items.isEmpty()) {
        BrandEmptyState(
            icon = Icons.Filled.Search,
            message = stringResource(AzkarR.string.azkar_search_empty),
            tokens = tokens,
        )
    } else {
        // Scrolls to and briefly highlights the item a content-reminder tap
        // named — a plain jump-to would land the reader on the right passage
        // with no sense of why THAT card, among a whole category of nearly
        // identical cards, is the one the notification meant. Keyed on the
        // item ids (not `items` itself) so this only re-runs when the actual
        // set of visible items changes, not on every recomposition.
        LaunchedEffect(items.map { it.id }, focusItemId) {
            val targetId = focusItemId ?: return@LaunchedEffect
            val index = items.indexOfFirst { it.id == targetId }
            if (index < 0) return@LaunchedEffect
            // The start-session row occupies index 0 when a category is shown.
            val listIndex = if (category != null) index + 1 else index
            listState.animateScrollToItem(listIndex)
            onHighlightChange(targetId)
            delay(2500)
            onHighlightChange(null)
        }
        LazyColumn(
            state = listState,
            verticalArrangement = Arrangement.spacedBy(12.dp),
            modifier = Modifier.fillMaxSize().padding(horizontal = 20.dp),
        ) {
            category?.let { cat ->
                item {
                    StartSessionRow(
                        category = cat,
                        isDone = isCompletedToday,
                        tokens = tokens,
                        onClick = { onStartSession(cat) },
                    )
                }
            }
            items(items, key = { it.id }) {
                AzkarEntryCard(it, tokens, locale, isHighlighted = it.id == highlightedItemId)
            }
            item { Spacer(Modifier.height(24.dp)) }
        }
    }
}

@Composable
private fun StartSessionRow(
    category: AzkarCategory,
    isDone: Boolean,
    tokens: ColorTokens,
    onClick: () -> Unit,
) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(10.dp),
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(14.dp))
            .background(tokens.primary.copy(alpha = 0.09f))
            .clickable(onClick = onClick)
            .padding(horizontal = 16.dp, vertical = 13.dp),
    ) {
        Icon(
            imageVector = if (isDone) Icons.Filled.Check else Icons.Filled.PlayArrow,
            contentDescription = null,
            tint = tokens.primary,
            modifier = Modifier.size(22.dp),
        )
        Text(
            text = stringResource(AzkarR.string.azkar_start_session),
            color = tokens.primary,
            fontWeight = FontWeight.SemiBold,
            style = MaterialTheme.typography.bodyMedium,
            modifier = Modifier.weight(1f),
        )
        Text(
            text = "${category.items.size}",
            color = tokens.onSurfaceSecondary,
            style = MaterialTheme.typography.labelMedium,
        )
    }
}

@Composable
private fun AzkarEntryCard(item: AzkarItem, tokens: ColorTokens, locale: String, isHighlighted: Boolean = false) {
    ArabicContentCard(
        arabic = item.arabicText,
        tokens = tokens,
        label = item.title,
        badgeText = RepeatCountLabel.text(item.repeatCount, Locale(locale)),
        modifier = if (isHighlighted) {
            Modifier.border(2.5.dp, tokens.primary, RoundedCornerShape(20.dp))
        } else {
            Modifier
        },
    )
}

// --- Du'a content ---

@Composable
private fun DuaContentSection(
    duas: List<Dua>,
    isSearching: Boolean,
    isFavorite: (String) -> Boolean,
    onToggleFavorite: (String) -> Unit,
    tokens: ColorTokens,
) {
    if (duas.isEmpty()) {
        // Distinguishing these matters: "no matches" and "this category is
        // empty" look identical otherwise, and the first is the user's doing
        // while the second is ours.
        BrandEmptyState(
            icon = Icons.Filled.Search,
            message = stringResource(if (isSearching) DuaR.string.dua_no_results else DuaR.string.dua_empty),
            tokens = tokens,
        )
    } else {
        LazyColumn(
            verticalArrangement = Arrangement.spacedBy(12.dp),
            modifier = Modifier.fillMaxSize().padding(horizontal = 20.dp),
        ) {
            items(duas, key = { it.id }) { dua ->
                ArabicContentCard(
                    arabic = dua.arabicText,
                    tokens = tokens,
                    label = dua.displayTitle,
                    accessory = {
                        FavoriteChip(
                            isFavorite = isFavorite(dua.id),
                            tokens = tokens,
                            onClick = { onToggleFavorite(dua.id) },
                        )
                    },
                )
            }
            item { Spacer(Modifier.height(24.dp)) }
        }
    }
}

@Composable
private fun FavoriteChip(isFavorite: Boolean, tokens: ColorTokens, onClick: () -> Unit) {
    val tint = if (isFavorite) tokens.accent else tokens.primary
    Box(
        modifier = Modifier
            .clip(CircleShape)
            .background(tint.copy(alpha = 0.10f))
            .clickable(onClick = onClick)
            .padding(horizontal = 11.dp, vertical = 7.dp),
    ) {
        Icon(
            imageVector = if (isFavorite) Icons.Filled.Favorite else Icons.Filled.FavoriteBorder,
            contentDescription = stringResource(
                if (isFavorite) DuaR.string.dua_unfavorite else DuaR.string.dua_favorite,
            ),
            tint = tint,
            modifier = Modifier.size(14.dp),
        )
    }
}

File: ios/App/Sources/Navigation/RemembranceScreen.swift
import AzkarFeature
import ContentKit
import DesignSystemKit
import DuaFeature
import SwiftUI

/// Azkar and Du'a merged into one screen behind a single category strip
/// (client direction, 2026-08-15) — replacing the earlier Azkar/Du'a
/// segmented control (2026-07-26) with one continuous row of chips spanning
/// both corpora, so the screen reads as one table of contents rather than
/// two tabs the reader has to know to switch between.
///
/// The two libraries still keep their own view models and their own session/
/// favourites logic; only the chip strip, search field, and card list are
/// unified here. A chip's *kind* (azkar category, dua category, or dua
/// favourites) decides which card style and per-item actions render below it.
struct RemembranceScreen: View {
    /// Which corpus opens by default — set by which deep link/tile got here
    /// (`fatwabot://azkar` vs `fatwabot://dua`). Both still land on this one
    /// screen; this only picks the initially-selected chip.
    enum Segment: String {
        case azkar, dua
    }

    private enum Selection: Equatable {
        case azkar(String)   // AzkarCategory.id
        case dua(String)     // DuaCategory.id
        case favorites
    }

    private let initial: Segment
    private let azkarViewModel: AzkarViewModel
    private let duaViewModel: DuaViewModel
    /// Which specific item to scroll to, from a content-reminder tap. Only
    /// azkar reminders exist today (`PlannedContentReminder.Kind` has no
    /// `.dua` case), so this only ever selects an azkar chip.
    private let focus: ContentFocus?

    @State private var selection: Selection?
    @State private var query = ""
    @State private var highlightedItemID: String?

    @Environment(\.colorScheme) private var colorScheme
    private var tokens: ColorTokens {
        colorScheme == .dark ? DesignTokens.bundledDefault.dark : DesignTokens.bundledDefault.light
    }

    init(
        initial: Segment, azkarViewModel: AzkarViewModel, duaViewModel: DuaViewModel,
        focus: ContentFocus? = nil
    ) {
        self.initial = initial
        self.azkarViewModel = azkarViewModel
        self.duaViewModel = duaViewModel
        self.focus = focus
    }

    private var azkarCategories: [AzkarCategory] { azkarViewModel.categories }
    private var duaCategories: [DuaCategory] { duaViewModel.categories }
    private var showsFavoritesChip: Bool { !duaViewModel.favoriteDuas.isEmpty }
    private var isLoading: Bool { azkarViewModel.isLoadingCategories || duaViewModel.isLoadingCategories }

    /// Resolves the stored selection to something guaranteed valid, falling
    /// back the same way each original screen did when the stored id goes
    /// stale: a content resync dropping a category, or unfavouriting the
    /// last dua while favourites is what's selected.
    private var resolvedSelection: Selection? {
        switch selection {
        case .azkar(let id) where azkarCategories.contains(where: { $0.id == id }):
            return selection
        case .dua(let id) where duaCategories.contains(where: { $0.id == id }):
            return selection
        case .favorites where showsFavoritesChip:
            return selection
        default:
            break
        }
        if let first = azkarCategories.first { return .azkar(first.id) }
        if let first = duaCategories.first { return .dua(first.id) }
        return nil
    }

    private var selectedAzkarCategory: AzkarCategory? {
        guard case .azkar(let id) = resolvedSelection else { return nil }
        return azkarCategories.first { $0.id == id }
    }

    private var selectedDuaCategory: DuaCategory? {
        guard case .dua(let id) = resolvedSelection else { return nil }
        return duaCategories.first { $0.id == id }
    }

    /// Search covers title, matn and source for azkar (before titles existed
    /// it was the only way most entries could be found by) and is scoped to
    /// the active category, unlike dua search below.
    private var visibleAzkarItems: [AzkarItem] {
        AzkarSearch.filter(selectedAzkarCategory?.items ?? [], query: query)
    }

    /// Dua search already spans every category once a query is entered
    /// (`DuaViewModel.searchResults`) — kept as-is rather than narrowed to
    /// the active chip.
    private var visibleDuas: [Dua] {
        if let results = duaViewModel.searchResults { return results }
        if resolvedSelection == .favorites { return duaViewModel.favoriteDuas }
        return selectedDuaCategory?.duas ?? []
    }

    var body: some View {
        VStack(spacing: 0) {
            searchField
            categoryChips
            content
        }
        .brandScreenBackground(tokens)
        .task {
            async let azkarLoad: Void = azkarViewModel.loadCategories(locale: "ar")
            async let duaLoad: Void = duaViewModel.loadCategories(locale: "ar")
            _ = await (azkarLoad, duaLoad)
            // Land on the category a reminder's item actually belongs to —
            // without this the item may not even be in `visibleAzkarItems`
            // yet, since that is filtered to whichever chip is selected.
            if let focusCategoryID = focus?.categorySlug {
                selection = .azkar(focusCategoryID)
            } else if selection == nil {
                selection = initial == .dua
                    ? duaCategories.first.map { .dua($0.id) }
                    : azkarCategories.first.map { .azkar($0.id) }
            }
        }
    }

    // MARK: - Header

    private var searchField: some View {
        HStack(spacing: 8) {
            Image(systemName: "magnifyingglass")
                .foregroundStyle(Color(hexToken: tokens.onSurfaceSecondary))
            TextField("azkar.search_placeholder", text: $query)
                .textFieldStyle(.plain)
                .foregroundStyle(Color(hexToken: tokens.onSurface))
                .autocorrectionDisabled()
                .submitLabel(.search)
            if !query.isEmpty {
                Button {
                    query = ""
                } label: {
                    Image(systemName: "xmark.circle.fill")
                        .foregroundStyle(Color(hexToken: tokens.onSurfaceSecondary))
                }
                .buttonStyle(.plain)
                .accessibilityLabel(Text("azkar.search_clear"))
            }
        }
        .padding(.horizontal, 14)
        .padding(.vertical, 11)
        .background(
            RoundedRectangle(cornerRadius: 14)
                .fill(Color(hexToken: tokens.surfaceElevated))
        )
        .padding(.horizontal, 20)
        .padding(.bottom, 12)
        .onChange(of: query) { _, newValue in
            // Dua's own search is what actually spans every dua category;
            // azkar filters locally via `visibleAzkarItems`.
            duaViewModel.searchQuery = newValue
        }
    }

    private var categoryChips: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 8) {
                ForEach(azkarCategories) { category in
                    chip(title: category.name, systemImage: nil, isSelected: resolvedSelection == .azkar(category.id)) {
                        select(.azkar(category.id))
                    }
                }
                if showsFavoritesChip {
                    chip(
                        title: String(localized: "dua.favorites"), systemImage: "heart.fill",
                        isSelected: resolvedSelection == .favorites
                    ) {
                        select(.favorites)
                    }
                }
                ForEach(duaCategories) { category in
                    chip(title: category.name, systemImage: nil, isSelected: resolvedSelection == .dua(category.id)) {
                        select(.dua(category.id))
                    }
                }
            }
            .padding(.horizontal, 20)
        }
        .padding(.bottom, 14)
    }

    private func select(_ newSelection: Selection) {
        selection = newSelection
        // A query from the previous chip almost never applies to the next
        // one, and leaving it set shows an empty list that reads as though
        // the category itself is empty.
        query = ""
        duaViewModel.searchQuery = ""
    }

    private func chip(title: String, systemImage: String?, isSelected: Bool, action: @escaping () -> Void) -> some View {
        Button(action: action) {
            HStack(spacing: 5) {
                if let systemImage {
                    Image(systemName: systemImage).font(.caption)
                }
                Text(verbatim: title)
                    .font(.subheadline.weight(isSelected ? .semibold : .regular))
            }
            .foregroundStyle(Color(hexToken: isSelected ? tokens.onPrimary : tokens.onSurface))
            .padding(.horizontal, 16)
            .padding(.vertical, 9)
            .background(
                Capsule().fill(
                    isSelected ? Color(hexToken: tokens.primary) : Color(hexToken: tokens.surfaceElevated)
                )
            )
        }
        .buttonStyle(.plain)
        .accessibilityAddTraits(isSelected ? [.isSelected] : [])
    }

    // MARK: - Body

    @ViewBuilder
    private var content: some View {
        if azkarCategories.isEmpty && duaCategories.isEmpty {
            if isLoading {
                ProgressView()
                    .tint(Color(hexToken: tokens.primary))
                    .frame(maxWidth: .infinity, maxHeight: .infinity)
            } else {
                BrandEmptyState(
                    systemImage: "book.closed", messageKey: "azkar.empty_state", tokens: tokens
                )
                .frame(maxHeight: .infinity)
            }
        } else {
            switch resolvedSelection {
            case .azkar:
                azkarContent
            case .dua, .favorites:
                duaContent
            case nil:
                ProgressView()
                    .tint(Color(hexToken: tokens.primary))
                    .frame(maxWidth: .infinity, maxHeight: .infinity)
            }
        }
    }

    // MARK: - Azkar content

    @ViewBuilder
    private var azkarContent: some View {
        if visibleAzkarItems.isEmpty {
            BrandEmptyState(
                systemImage: "magnifyingglass", messageKey: "azkar.search_empty", tokens: tokens
            )
            .frame(maxHeight: .infinity)
        } else {
            ScrollViewReader { proxy in
                ScrollView {
                    LazyVStack(spacing: 12) {
                        if let category = selectedAzkarCategory {
                            startSessionRow(category)
                        }
                        ForEach(visibleAzkarItems) { item in
                            azkarCard(item).id(item.id)
                        }
                    }
                    .padding(.horizontal, 20)
                    .padding(.bottom, 24)
                }
                // Runs once the focused item actually appears in the list —
                // it isn't there on the first pass while `.task` above is
                // still loading and switching to the right category.
                .onChange(of: visibleAzkarItems.map(\.id)) { _, ids in
                    guard let focusItemID = focus?.contentID, ids.contains(focusItemID) else { return }
                    scrollToFocus(focusItemID, proxy: proxy)
                }
                .onAppear {
                    guard let focusItemID = focus?.contentID,
                          visibleAzkarItems.contains(where: { $0.id == focusItemID }) else { return }
                    scrollToFocus(focusItemID, proxy: proxy)
                }
            }
        }
    }

    /// Scrolls to and briefly highlights the item a content-reminder tap
    /// named — a plain jump-to would land the reader on the right passage
    /// with no sense of why THAT card, among a whole category of nearly
    /// identical cards, is the one the notification meant.
    private func scrollToFocus(_ id: String, proxy: ScrollViewProxy) {
        withAnimation { proxy.scrollTo(id, anchor: .center) }
        withAnimation { highlightedItemID = id }
        Task {
            try? await Task.sleep(nanoseconds: 2_500_000_000)
            withAnimation { highlightedItemID = nil }
        }
    }

    /// The way into the counting session, kept at the top of the list rather
    /// than on every card: starting a session is a per-category action, and
    /// repeating it on every card would imply each one starts its own.
    private func startSessionRow(_ category: AzkarCategory) -> some View {
        NavigationLink {
            AzkarSessionScreen(viewModel: azkarViewModel, category: category)
        } label: {
            HStack(spacing: 10) {
                Image(systemName: azkarViewModel.isCompletedToday(category.id)
                      ? "checkmark.seal.fill" : "play.circle.fill")
                    .font(.title3)
                Text("azkar.start_session")
                    .font(.subheadline.weight(.semibold))
                Spacer(minLength: 0)
                Text("\(category.items.count)")
                    .font(.caption.monospacedDigit())
                    .foregroundStyle(Color(hexToken: tokens.onSurfaceSecondary))
            }
            .foregroundStyle(Color(hexToken: tokens.primary))
            .padding(.horizontal, 16)
            .padding(.vertical, 13)
            .background(
                RoundedRectangle(cornerRadius: 14)
                    .fill(Color(hexToken: tokens.primary).opacity(0.09))
            )
        }
        .buttonStyle(.plain)
    }

    private func azkarCard(_ item: AzkarItem) -> some View {
        ArabicContentCard(
            label: item.title,
            badgeText: RepeatCountLabel.text(item.repeatCount, locale: Locale(identifier: "ar")),
            arabic: item.arabicText,
            tokens: tokens
        )
        .overlay(
            RoundedRectangle(cornerRadius: 20, style: .continuous)
                .stroke(Color(hexToken: tokens.primary), lineWidth: highlightedItemID == item.id ? 2.5 : 0)
        )
    }

    // MARK: - Du'a content

    @ViewBuilder
    private var duaContent: some View {
        if visibleDuas.isEmpty {
            // Distinguishing these matters: "no matches" and "this category is
            // empty" look identical otherwise, and the first is the user's
            // doing while the second is ours.
            BrandEmptyState(
                systemImage: "magnifyingglass",
                messageKey: duaViewModel.searchResults == nil ? "dua.empty_state" : "dua.search_no_results",
                tokens: tokens
            )
            .frame(maxHeight: .infinity)
        } else {
            ScrollView {
                LazyVStack(spacing: 12) {
                    ForEach(visibleDuas) { dua in
                        duaCard(dua)
                    }
                }
                .padding(.horizontal, 20)
                .padding(.bottom, 24)
            }
        }
    }

    private func duaCard(_ dua: Dua) -> some View {
        ArabicContentCard(
            label: dua.displayTitle,
            arabic: dua.arabicText,
            tokens: tokens
        ) {
            favoriteButton(dua)
        }
    }

    private func favoriteButton(_ dua: Dua) -> some View {
        let isFavorite = duaViewModel.isFavorite(dua.id)
        return Button {
            duaViewModel.toggleFavorite(dua.id)
        } label: {
            Image(systemName: isFavorite ? "heart.fill" : "heart")
                .font(.caption.weight(.medium))
                .foregroundStyle(Color(hexToken: isFavorite ? tokens.accent : tokens.primary))
                .padding(.horizontal, 11)
                .padding(.vertical, 7)
                .background(
                    Capsule().fill(
                        Color(hexToken: isFavorite ? tokens.accent : tokens.primary).opacity(0.10)
                    )
                )
        }
        .buttonStyle(.plain)
        .accessibilityLabel(Text(isFavorite ? "dua.unfavorite" : "dua.favorite"))
    }
}
